### PR TITLE
Gallery with examples

### PR DESCRIPTION
**Pull request by reneeotten**
#### Description
This PR adds a gallery with examples using ```sphinx-gallery``` (see #544). It's only as a "proof-of-principle": I have added the gallery to the HTML documentation and added only one (still to-be-improved) example. You can test it by [checking out the PR locally](https://help.github.com/en/articles/checking-out-pull-requests-locally), install the ```sphinx-gallery``` package and run ```make``` in the ```doc``` directory.

Before it can be merged (certainly after #573) we'll need to go through the examples, add the correct header, update them to follow recommended Python coding (style) rules, and add some descriptive text/explanations. Also, we should invite people (perhaps on the mailinglist) to submit self-contained, real-life examples that show how ```lmfit``` can be used.

###### Type of Changes

- [ ] Bug fix
- [ ] New feature
- [ ] Refactoring / maintenance
- [x] Documentation / examples

###### Verification 
Have you
- [x] referenced existing Issue and/or provided relevant link to mailing list?
- [x] squashed/minimized your commits and written descriptive commit messages?

### CODE REVIEW

**Comment by newville** [comment]
@reneeotten I got an exception of 

```
cp sphinx/ext_mathjax.py extensions.py
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v1.8.5
loading pickled environment... done
generating gallery...

Exception occurred:
  File "/home/newville/anaconda3/lib/python3.7/site-packages/sphinx_gallery/py_source_parser.py", line 103, in get_docstring_and_rest
    .format(filename))
ValueError: Could not find docstring in file "/home/newville/Codes/lmfit-py/examples/a.py". A docstring is required by sphinx-gallery unless the file is ignored by "ignore_pattern"
The full traceback has been saved in /tmp/sphinx-err-dtlhwzaa.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:22: html] Error 2
```

That's with Python 3.7.3, sphinx 1.8.5, sphinx_gallery 0.4.0 on Linux.  I have not looked into any further yet.   But - this is going to be great!

**Comment by reneeotten** [comment]
@newville it looks like you have locally a file called ```a.py``` in your "examples" directory. Currently, all the files present in that directory will be processed by ```sphinx-gallery``` with the exception of the ```doc_*.py``` ones that are used in the documentation. 

For the files that are used in the gallery they will all need to have a correct header / docstring that will be used as name and description for the entry in the galler, as shown below:
```
"""
Example name
======================

Example description
"""
```

The error message you're seeing is because the ```a.py``` does not contain this header, but you're (probably unknowingly) asking it to be included.

**Comment by newville** [comment]
@reneeotten Ah, yes, thanks!  That makes sense and now works and looks good.  Cool! 

Should we make a clearer distinction between the examples that are in-lined into the main docs and those intended for the gallery? I don't have a strong preference on how that would be organized.     

I'd be OK with merging #573 and this and then organizing examples, but if you have ideas for doing some of that before merging, that's OK with me too.

**Comment by reneeotten** [comment]
> Should we make a clearer distinction between the examples that are in-lined into the main docs ? and those intended for the gallery? I don't have a strong preference on how that would be organized.

we have had that distinction already for a while: files starting with ```doc_``` are inlined/used in the main documentation, whereas all others are not. I would prefer to keep it like and that's what I have done in this PR: files that start with ```doc_``` are not processed for the gallery whereas all others are.

If you're happy with #573, feel free to merge that one; I will then rebase this PR (that touches some of the same files). I will likely not have much time (in the near future) to do any real work on improving the examples (hopefully others will step up here). Having said that, what I can do *before* merging this PR is to quickly go over the current examples and at least have most of them included in the gallery. There isn't much point to add a gallery and have no examples included ;) That would keep us pretty much in the same situation as now, but with the added benefit that examples are clearly visible on the website (instead of someone needing to go and dig through ```conda``` or ```pip``` installation locations. From there someone can start the clean-up and improvement phase - how about that?

**Comment by newville** [comment]
@reneeotten  Sorry, it looks like maybe I should have merged this before #575 to avoid conflicts.  I'm happy to merge this and it looks like conflicts are small -- I think I could resolve these if you don't have the time. 

Let me know if I should undo the merge of #575.....

**Comment by reneeotten** [comment]
> @reneeotten Sorry, it looks like maybe I should have merged this before #575 to avoid conflicts. I'm happy to merge this and it looks like conflicts are small -- I think I could resolve these if you don't have the time.
> 
> Let me know if I should undo the merge of #575.....

This is fine; the order wouldn't matter you would get conflicts in the "second" PR anyway as they touch the sames file. I have fixed them locally already. 

I will work a bit more on the examples so that we incorporate whatever we have right now into the gallery and will push the changes to this PR. So please don't merge anything yet - to be continued!

**Comment by reneeotten** [comment]
@newville can you take a look at the current files in the ```examples``` directory and give your opinion on which ones should stay or go? I think some of them can be merged because of similarity; the NIST related ones can go as far as I am concerned. I will finish up making all of them render in the gallery and do a little (easy-stuff) clean-up.

I am currently having difficulties with the example ```fit_with_analytic_jacobian.py```: it does "finish" in less iterations but the result are terrible... What's the purpose of this example and am I missing something obvious here? 

To be honest though, I think it would pretty much require to start from scratch with carefully selected, user-guide style, examples to be really useful for new users. But that's not something that will happen overnight (or likely ever by me)... Perhaps we should ask on the mailinglist for users to submit some of their examples for the gallery or, even better, try to get someone interested to work on this topic and ideally the documentation part?

**Comment by newville** [comment]
@reneeotten As far as I'm concerned, all the examples that aren't `doc_*` can be purged, and many of the `doc_*` examples can be revisited and/or heavily purged -- your discretion.  If any of them aren't working, just toss them.   The code isn't completely lost and we can revisit any of these if needed. 
 
The NIST examples are sort of an interesting set of examples, but they're very old and were mostly for investigating in the early days of lmfit how the different solvers performed.  It might make an interesting gallery entry or Jupyter notebook, but can almost certainly benefit from starting over too. I'll take that as my responsibility.

In short, I completely agree that starting over is a fine idea.  Soliciting ideas from the mailing list is a fine idea.  Here are a few near-FAQ fitting categories that might be nice gallery entries and/or Jupyter examples are

- fitting multiple datasets
- fitting exponentially decaying data (fitting in log space)
- inequality constraints
- brute force methods
- advanced confidence intervals

I'm not saying these are needed now.  Maybe these are candidates for requests for inputs from interested users or inspirations for the sorts of examples that might help other users.

**Comment by reneeotten** [comment]
@newville sorry for the delay, but I think this is what we could merge for now. The code in this PR does the following:

1. documentation examples
The ```doc_*.py``` files are temporarily copied to ```/examples/documentation``` and a proper header is added. All these examples will show up in a sub-gallery.

2. other examples
I have gone through the examples, combined some of them, removed a few (IMHO) useless ones, and as discussed above got rid of the ```NIST``` examples as well. The idea here is that all files should be in "plain" Python code that contain comments/descriptive text in rST; they will be processed by ```sphinx-gallery``` and included as ```Jupyter``` notebooks. 
There are scenarios: all files that start with ```example_``` are executed, whereas all other are only parsed (this is, for example, useful when examples would take a long time to run).

As I said before, I think most of the (non-documentation) examples would benefit greatly from someone updating/improving them, and your list of examples above would be certainly good to add. I consider this a good starting point that shows how it could work and try to improve from here.

**Comment by reneeotten** [comment]
ah, Travis complains as it has difficulties in building the documentation: it appears unable to find the data files.... It certainly works locally, so feel free to try already yourself for now. I will take a look at fixing this for Travis.

**Comment by newville** [comment]
@reneeotten  Thanks!  I did run into a problem with sphinx complaining about a docstring, but I think this was a sphinx version issue.   I "pip upgraded" and it went away.  

This was a ton of work. Like you say, this is now ready for more work, but it's already great. 

I'll just merge this and we can go from here!   Do you think we should push the gallery to be more easily found from the main doc page?

**Comment by reneeotten** [comment]
> Do you think we should push the gallery to be more easily found from the main doc page?

I did change ```doc/sphinx/theme/lmfitdoc/layout.html``` already so that there now is an ```examples``` entry in the menu-bar, next to ```constraints```. I would think that's enough, but if you feel like something else is needed please do!
